processor/k8sprocessor/processor.go

```diff
@@ -92,56 +92,31 @@ func (kp *kubernetesprocessor) ProcessMetrics(ctx context.Context, md pdata.Metr
 func (kp *kubernetesprocessor) ProcessLogs(ctx context.Context, ld pdata.Logs) (pdata.Logs, error) {
 	rl := ld.ResourceLogs()
 	for i := 0; i < rl.Len(); i++ {
-		kp.processResource(ctx, rl.At(i).Resource(), k8sIPFromAttributes())
+		kp.processResource(ctx, rl.At(i).Resource())
 	}
 
 	return ld, nil
 }
 
-func (kp *kubernetesprocessor) processResource(ctx context.Context, resource pdata.Resource, attributeExtractors ...ipExtractor) {
-	var podIP string
+func (kp *kubernetesprocessor) processResource(ctx context.Context, resource pdata.Resource) {
 
-	for _, extractor := range attributeExtractors {
-		podIP = extractor(resource.Attributes())
-		if podIP != "" {
-			break
-		}
-	}
-
-	// Check if the receiver detected client IP.
-	if podIP == "" {
-		if c, ok := client.FromContext(ctx); ok {
-			podIP = c.IP
-		}
-	}
-
-	// If IP is still not available by this point, nothing can be tagged here. Return.
-	if podIP == "" {
+	podIdentifiers := k8sPodAssociationFromAttributes(ctx, resource.Attributes(), kp.podAssociations)
+	if len(podIdentifiers) == 0 {
 		return
 	}
-
-	resource.Attributes().InsertString(k8sIPLabelName, podIP)
-
-	// Don't invoke any k8s client functionality in passthrough mode.
-	// Just tag the IP and forward the batch.
-	if kp.passthroughMode {
-		return
-	}
-
-	// add k8s tags to resource
-	attrsToAdd := kp.getAttributesForPodIP(podIP)
-	if len(attrsToAdd) == 0 {
-		return
-	}
-
-	attrs := resource.Attributes()
-	for k, v := range attrsToAdd {
-		attrs.InsertString(k, v)
+	for k, v := range podIdentifiers {
```

Comment:
If I understand this loop correctly it may find the same pod more than once and add try to insert `attrsToAdd` more than once (if len(podIdentifiers)>1) which is unnecessary work since the second and subsequent insertion attempts are noop.

Worse, it may find different pods if `podIdentifiers` happens to identify different pods for whatever reason and in that case it will insert attributes of different pods into this `resource`.

Shouldn't this process stop once the first match happens and the pod is found (as it is described in the docs)? The problem is that breaking this loop is not going to achieve that since the order of iteration here is different from the order of rules in `kp.podAssociations`.

Generally, I don't understand why instead of having `k8sPodAssociationFromAttributes` find all identifiers and then iterate over all identifiers here, we don't iterate over `kp.podAssociations` and break after the first match. That I think would result in the behavior that matches the docs.